docs/specs/priority-queue.md

```diff
@@ -1,70 +1,99 @@
+
+## The proposed classes
+
+There are actually two PriorityQueue classes, one to support each scenario. PriorityQueue<TItem, TPriority> supports the *extrinsic priority* scenario, and PrioritySet<T> supports the *intrinsic priority* scenario. 
+
+## `PrioritySet<T>` (for intrinsic priority)
+
+`PrioritySet<T>` is a set of items with intrinsic priorities. It must be initialized with a 'priority comparer' which understands how to compare the intrinsic priorities of each item.
+
+It differs from `SortedSet<T>` because it allows separate items to have equal priorities, and because it does not maintain strict ordering at all times, and because it allows that priorities are mutable.
+
+If you provide `DefaultComparer<T>` as your priority comparer, and never mutate priorities, you will get similar behavior to `SortedSet<T>`, where priorities are the items and priorities are all distinct.
 
 ```csharp
-public class PriorityQueue<T> : IQueue<T>
+public class PrioritySet<T> : ISet<T>, ICollection<T>, IEnumerable<T>, IEnumerable, 
 {
-    public PriorityQueue();
-    public PriorityQueue(IComparer<T> comparer);
-    public PriorityQueue(IEnumerable<T> collection);
-    public PriorityQueue(IEnumerable<T> collection, IComparer<T> comparer);
+    public PriorityQueue(IComparer<T> priorityComparer);
+    public PriorityQueue(IComparer<T> priorityComparer, IEqualityComparer<T> equalityComparer);
+
+    public PriorityQueue(IEnumerable<T> collection, IComparer<T> priorityComparer);
+    public PriorityQueue(IEnumerable<T> collection, IComparer<T> priorityComparer, IEqualityComparer<T> equalityComparer);
+
+    public IComparer<T> PriorityComparer { get; } // compares items using .CompareTo() to see which is higher priority
+    public IComparer<T> EqualityComparer { get; } // comapres items using .GetHashCode() and .Equals() to see if they are the same as existing items in the collection
 
-    public IComparer<T> Comparer { get; }
     public int Count { get; }
+    public bool IsReadOnly { get; }
 
-    public bool IsEmpty { get; }
+    public bool Add(T item);
     public void Clear();
-    public bool Contains(T element);
-
-    public void Enqueue(T element);
+    public bool Contains(T item);
+    public void CopyTo(T[] item, int index);
+    public bool Remove(T item);
 
     public T Peek();
     public T Dequeue();
-    public bool Remove(T element);
 
-    public bool TryPeek(out T element);
-    public bool TryDequeue(out T element);
+    public bool TryPeek(out T item);
+    public bool TryDequeue(out T item);
+
+    public void NotifyPriorityChanged(T item);
+    public void NotifyPrioritiesChanged(IEnumerable<T> items);
 
     public IEnumerator<T> GetEnumerator();
-    IEnumerator IEnumerable.GetEnumerator();
+
+    // ommitted: declaration of various other ICollection, IEnumerable, and ISet interface members (ExceptWith, Overlaps, IsProperSubsetOf, IsSubsetOf, IsSuperSetOf, IntersectWith, UnionWith, etc.)
```

Comment:
exactly, how would these work in the context of a priority queue?

Reply:
@nietras I imagine that they would work just as if the priority queue is consider as 'a set of elements... which just happen to have priorities' e.g. IntersectWith removes all elements from the priority queue which are not part of the target collection.

Reply:
@TimLovellSmith right like `SortedSet` and `HashSet`, I understand that, but does that make sense in this case? There might not be an as efficient way to do those? Depending on implementation. I guess the question is what use case are there for set operations on a priority queue? Perhaps there are valid ones, I just wondered why? :)

Reply:
@nietras Maybe the best answer here is that we don't really need PrioritySet at all. There's no PrioritySet scenario that can't instead be implemented with external priority PriorityQueue, and the downsides of having mutable intrinsic priorities are a huge pit of failure.